File: frontend/types/history.ts
export interface HistoryNode {
  id: number;
  title: string;
  url: string;
  prev: number | null;
  next: number | null;
}

export interface HistoryResponse {
  head: number | null;
  tail: number | null;
  current: number | null;
  nodes: HistoryNode[];
}

export type OperationKind = "INSERT" | "BACKWARD" | "FORWARD" | "DELETE" | "SEARCH" | "CLEAR";

export interface OperationLogEntry {
  id: number;
  kind: OperationKind;
  label: string;
  detail: string;
  time: string;
}

File: backend/models/history.py
from pydantic import BaseModel
from typing import List, Optional


class VisitRequest(BaseModel):
    url: str
    title: str


class NodeResponse(BaseModel):
    id: int
    title: str
    url: str
    prev: Optional[int] = None
    next: Optional[int] = None


class HistoryResponse(BaseModel):
    head: Optional[int] = None
    tail: Optional[int] = None
    current: Optional[int] = None
    nodes: List[NodeResponse]


class MessageResponse(BaseModel):
    message: str